Add custom source suffix and HTTP Headers for custom urls with authentication support via env var interpolation

solvingj opened the pull request:
This was intended to be included in https://github.com/codrsquad/portable-python/pull/59
The use case is pretty straightforward and it is working

zsimic reviewed (approved):
To satisfy the linter, let's bump McCabe allowed complexity to 18, I'll refactor this to come back to simpler function later.

Change the setting here: https://github.com/codrsquad/portable-python/blob/main/pyproject.toml#L66 from 14 to 18

solvingj commented:
Ok, fixed that, and refactored a bit, added another url fix, and added another forgotten local change of the suffix override for github release assets.

zsimic commented on src/portable_python/__init__.py:
```diff
@@ -632,8 +654,15 @@ def compile(self):
+                    basename = f"{self.m_name}-{self.version}.{suffix}"
+                else:
+                    # Split on '#' for urls that include a checksum, such as #sha256=... fragment
+                    basename = runez.basename(self.url, extension_marker="#")
```
I think we need to split first on `#`, and then look for known extension

solvingj replied:
Makes sense, please feel free to edit if you are available

zsimic replied:
Corrected and added test to observe that a `#sha256=...` doesn't fool this into thinking url has no known extension.

zsimic commented:
Thank you so much for the contrib once again!!

zsimic commented:
Released as v1.9.9
